### PR TITLE
Perform topic check on publish

### PR DESCRIPTION
**Pull request by ntoonio**
I was asked (https://github.com/beerfactory/hbmqtt/pull/195#issuecomment-798745711, https://github.com/Yakifo/amqtt/issues/7) to add this feature to this fork since this didn't get merged before `hbmqtt` was deprecated. I hope I did this in the right way.

This pull request consists of two pull requests, the latter built on top of the first;

- https://github.com/beerfactory/hbmqtt/pull/190 applies plugin topic checks to publish commands as well (@jacoposartini)
- https://github.com/beerfactory/hbmqtt/pull/195 lets the plugin differentiate between publish and subscribe calls by adding a `command` kwarg (me, @ntoonio)

Can be used as:
```
class CustomTopicPlugin(BaseTopicPlugin):
    def __init__(self, context):
        super().__init__(context)

    async def topic_filtering(self, *args, **kwargs):
        publish = kwargs.get("command") == 1
        print("This is a " + ("publish" if publish else "subscribe") + " command")
        return True
```

### CODE REVIEW

**Comment by HerrMuellerluedenscheid** [comment]
@ntoonio Thanks for opening the MR! Can you run `black` on the modified files? Should work with the defaults.

**Comment by HerrMuellerluedenscheid** [comment]
@ntoonio Thanks for fixing the linter! Can you add a `pytest` that reflect the changes?

Also, I think this is quite a lot of commits with repetitive commit messages please consider an interactive rebase to squash the commits.

**Comment by ntoonio** [comment]
Will be adding more test for actually running the plugins some other time.

How should I go about merging the commits `520e7ea` to `c505741`? Using `rebase -i` i have to fix merge conflicts with the current HEAD for every commit... I tried checking out to `c505741` and then doing `rebase` which worked great, but how do i save those changes? When i went back the commits hadn't been sqashed

**Comment by FlorianLudwig** [comment]
@ntoonio 

Here a few tips on how to use `rebase -i`

# HOWTO: git rebase

`git rebase -i` is a powerful tool which can accomplish different task:

1. rebasing 
2. linearizing the history
3. changing history (squash, reorder, delete or alter commits)

Only do one of those at a time!
Usually, if a `git rebase -i` goes wrong it is because multiple of these things were done at the same time - without the user realizing it.
To avoid this always:

1. Take a good look at your git history before using `rebase`
2. Be sure what you want to accomplish
3. Never use `git rebase -i TARGET` without specifying a target


## rebasing

1. Make sure there are no merge commits in your current branch
2. Do not use `-i` at all

## linearizing the history

1. Find a commit before the branches diverged
2. `git rebase -i COMMIT`
3. If any commits were on both branches, delete one of them
4. Ensure the order of your commits is as expected

## changing history

Ensure:

1. No merge commits in the history you are about to change
2. You are not actually rebasing

One easy way to figure out if you are doing any of those two things is to run your `git rebase -i TARGET` command without specifying any changes.
If a conflict arises, you were actually rebasing or had a merge in the history.

**Comment by FlorianLudwig** [comment]
@ntoonio if you run `git rebase -i origin/master` you should drop the merge commit (`18aecd5`). Merging and rebasing in the same workflow doesn't work well.

**Comment by FlorianLudwig** [comment]
@ntoonio hey!

We moved to poetry - hope this makes it a bit easier to setup a development environment.

Could you check: https://github.com/Yakifo/amqtt/blob/master/contributing.md

And let me know if that is helpful? Or how we could improve it?

**Comment by HerrMuellerluedenscheid** [comment]
> Will be adding more test for actually running the plugins some other time.
> 
> How should I go about merging the commits `520e7ea` to `c505741`? Using `rebase -i` i have to fix merge conflicts with the current HEAD for every commit... I tried checking out to `c505741` and then doing `rebase` which worked great, but how do i save those changes? When i went back the commits hadn't been squashed
> 
> Also... is there any way yo do the `Linting with flake8 and black` test locally? I'm running black on all files and `pytest` but the tests on github still fails.

`black` doesn't seem to be a problem but `flake8`: https://github.com/Yakifo/amqtt/pull/42/checks?check_run_id=2152892451

You can run `flake8` manually:
```
flake8 . --count --statistics --show-source --ignore=E501,W503,W605,E722
```

Alternatively, you can use https://pre-commit.com/. you might have to rebase on master. We added a `pre-commit.yaml` in the meantime. To manually run the style checks:
```
pre-commit run --all-files
```

File: hbmqtt/plugins/topic_checking.py [hunk 1]
```diff
@@ -24,16 +24,24 @@ def topic_filtering(self, *args, **kwargs):
...
             session = kwargs.get("session", None)
             topic = kwargs.get("topic", None)
+            command = kwargs.get("command")
             if session.username and session.username == "admin":
                 return True
-            if topic and topic in self._taboo:
+            if (
+                topic
+                and topic in self._taboo
+                and (self._taboo_command == None or self._taboo_command == command)
```
Comment (HerrMuellerluedenscheid) [inline hbmqtt/plugins/topic_checking.py:43]: compare against `None` with `is` not `==`.

File: tests/plugins/test_topic_checking.py [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+        ret = self.loop.run_until_complete(
+            taboo_topic_plugin.topic_filtering(topic="allowed", command=0, session=s)
+        )
+        context.config["topic-check"]["taboo_command"] = None
+        taboo_topic_plugin = TopicTabooPlugin(context)
+        assert ret
+        ret = self.loop.run_until_complete(
+            taboo_topic_plugin.topic_filtering(topic="prohibited", command=1, session=s)
+        )
+        assert ret == False
```
Comment (HerrMuellerluedenscheid) [inline tests/plugins/test_topic_checking.py:61]: compare against `False` with `is` not `==`.

File: tests/plugins/test_topic_checking.py [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+        ret = self.loop.run_until_complete(
+            taboo_topic_plugin.topic_filtering(topic="prohibited", command=0, session=s)
+        )
+        assert ret == False
```
Comment (HerrMuellerluedenscheid) [inline tests/plugins/test_topic_checking.py:65]: Fix assort as above

**Review by HerrMuellerluedenscheid — CHANGES_REQUESTED** [review]

File: tests/plugins/test_topic_checking.py [hunk 4]
```diff
@@ -0,0 +1,72 @@
+# Copyright (c) 2015 Nicolas JOUANIN
+#
+# See the file license.txt for copying permission.
+
+import unittest
+import logging
+import os
```
Comment (HerrMuellerluedenscheid) [inline tests/plugins/test_topic_checking.py:7]: Remove unused import

**Comment by ntoonio** [comment]
I just can't figure out how I would go about squashing those commits... I tried to drop the merge commits, but that removed to commits before that, that I wanted to squash. I think i did the merging backwards. Instead of merging the new amqtt into my branch. I cloned amqtt and merged my branch. It seemed easy at the time but not I think it's just resulting in problems. Would it be easier to redo in a new PR? Feeling kinda stuck right now

**Comment by HerrMuellerluedenscheid** [comment]
> I just can't figure out how I would go about squashing those commits... I tried to drop the merge commits, but that removed to commits before that, that I wanted to squash. I think i did the merging backwards. Instead of merging the new amqtt into my branch. I cloned amqtt and merged my branch. It seemed easy at the time but not I think it's just resulting in problems. Would it be easier to redo in a new PR? Feeling kinda stuck right now

I gave it a try running an interactive rebase on this branch. I also ran into hard to resolve merge conflicts. I'll cherry-pick or do a plain copy-paste some day.

**Comment by FlorianLudwig** [comment]
closing this in favor of https://github.com/Yakifo/amqtt/pull/66

**Pull request closed without merging**